design-proposals/vm-pool.md

```diff
@@ -0,0 +1,242 @@
+* The VMConfig has a different status section from VM
+* A VMConfig’s name does not correlate in any way to a VM generated from the VMConfig
+* In a VMConfig, all naming resulting in newly created objects are treated as generated names in practice. For example, the name of a DataVolumeTemplate in a VMConfig.Spec.DataVolumeTemplates list will be treated as a generated name when VMs are spawned using the config. This ensures all VMs derived from a VMConfig are unique.
+
+**Similarities**
+* VMConfig and VM object’s spec structure are identical
+
+## VirtualMachinePool (VMPool) API
+
+The VMPool API represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
+
+The VMPools spec contains the following tunings and values
+
+* **VirtualMachineConfigRef** - (Required) A structure that points to a VMConfig object
+	* **Name** - (Required) The name of the VMConfig to assign to this VMPool. Assumes the object type is VirtualMachineConfig and that the object exists in the same namespace as the pool.
+* **Replicas** - (Required) An integer representing the desired number of VM replicas
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
+* **NameGeneration** - (Optional) Specifies how objects within a pool have their names generated
+	* **VMPrefix** - (Defaults to VMPool.Name) - String representing the prefix to use for when naming VMs.
+	* **AppendPostfixToSecretReferences** - (default false) Boolean that indicates if VM’s unique postfix should be appended to references to Secrets in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+	* **AppendPostfixToConfigMapReferences** - (default false) Boolean that indicates if VM’s unique postfix should be appended to ConfigMap references in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+
+* **UpdateStrategy** - (Optional) Specifies how the VMPool controller manages updating VMs within a VMPool
+	* **Unmanaged** - No automation during updates. The VM is never touched after creation. Users manually update individual VMs in a pool.
+	* **Opportunistic** - Opportunistic update of VMs which are in a halted state.
+	* **Proactive** - (Default) Proactive update by forcing VMs to restart during update.
+		* **SelectionPolicy** - (Optional) (Defaults to "random" base policy when no SelectionPolicy is configured) The priority in which VM instances are selected for proactive scale-in
+			* **OrderedPolicies** - (Optional) Ordered list of selection policies. Policies include [LabelSelector|NodeSelector]
```

Comment:
From SIG-scale: During Update and scale-in, there could be VMIs that are not `Running` yet.  There could be an additional policy here to target VMIs that are not `Running` first, before deleting VMIs that are (and vice-versa) - it would improve upgrade speed.